From 6a544e0456414de849aa2271d20d6ffa82c9812b Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Fri, 8 Dec 2017 16:37:28 +0000
Subject: [PATCH 5/6] Implement the new remove-orphans command

This command will remove any packages that were automatically installed
and have been identified as orphaned by nature of having no reverse
dependencies outside of the orphan list.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 pisi/api.py               |  5 +++++
 pisi/cli/pisicli.py       |  1 +
 pisi/cli/removeorphans.py | 52 +++++++++++++++++++++++++++++++++++++++++++++++
 pisi/operations/remove.py |  8 ++++++--
 4 files changed, 64 insertions(+), 2 deletions(-)
 create mode 100644 pisi/cli/removeorphans.py

diff --git a/pisi/api.py b/pisi/api.py
index 9f60c61..56c57cb 100644
--- a/pisi/api.py
+++ b/pisi/api.py
@@ -441,6 +441,11 @@ def autoremove(packages, ignore_dependency=False, ignore_safety=False):
     pisi.db.historydb.HistoryDB().create_history("remove")
     return pisi.operations.remove.remove(packages, ignore_dependency, ignore_safety, autoremove=True)
 
+@locked
+def remove_orphans(ignore_dependency=False, ignore_safety=False):
+    pisi.db.historydb.HistoryDB().create_history("remove")
+    return pisi.operations.remove.remove_orphans(ignore_dependency, ignore_safety)
+
 @locked
 def install(packages, reinstall=False, ignore_file_conflicts=False, ignore_package_conflicts=False):
     """
diff --git a/pisi/cli/pisicli.py b/pisi/cli/pisicli.py
index 50cf2a1..7c5fcde 100644
--- a/pisi/cli/pisicli.py
+++ b/pisi/cli/pisicli.py
@@ -45,6 +45,7 @@ import pisi.cli.listsources
 import pisi.cli.listupgrades
 import pisi.cli.rebuilddb
 import pisi.cli.remove
+import pisi.cli.removeorphans
 import pisi.cli.removerepo
 import pisi.cli.enablerepo
 import pisi.cli.disablerepo
diff --git a/pisi/cli/removeorphans.py b/pisi/cli/removeorphans.py
new file mode 100644
index 0000000..6b049b1
--- /dev/null
+++ b/pisi/cli/removeorphans.py
@@ -0,0 +1,52 @@
+# -*- coding:utf-8 -*-
+#
+# Copyright (C) 2005 - 2007, TUBITAK/UEKAE
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free
+# Software Foundation; either version 2 of the License, or (at your option)
+# any later version.
+#
+# Please read the COPYING file.
+#
+
+import optparse
+
+import gettext
+__trans = gettext.translation('pisi', fallback=True)
+_ = __trans.ugettext
+
+import pisi.cli.command as command
+import pisi.context as ctx
+import pisi.api
+import pisi.db
+
+class RemoveOrphans(command.PackageOp):
+    __doc__ = _("""Remove orphaned packages
+
+Usage: remove-orphans
+
+Remove any unused orphan packages from the system that were automatically
+installed as a dependency of another package.
+
+Only packages that have no reverse dependencies outside of the automatically
+installed list will be removed.
+""")
+    __metaclass__ = command.autocommand
+
+    def __init__(self, args):
+        super(RemoveOrphans, self).__init__(args)
+
+    name = ("remove-orphans", "rmo")
+
+    def options(self):
+        group = optparse.OptionGroup(self.parser, _("remove-orphans options"))
+        super(RemoveOrphans, self).options(group)
+        group.add_option("--purge", action="store_true",
+                     default=False, help=_("Removes everything including changed config files of the package"))
+        self.parser.add_option_group(group)
+
+    def run(self):
+        self.init()
+
+        pisi.api.remove_orphans()
diff --git a/pisi/operations/remove.py b/pisi/operations/remove.py
index 0a1cdf1..548a920 100644
--- a/pisi/operations/remove.py
+++ b/pisi/operations/remove.py
@@ -24,7 +24,7 @@ import pisi.util as util
 import pisi.ui as ui
 import pisi.db
 
-def remove(A, ignore_dep = False, ignore_safety = False, autoremove = False):
+def remove(A, ignore_dep = False, ignore_safety = False, autoremove = False, force_prompt = False):
     """remove set A of packages from system (A is a list of package names)"""
 
     componentdb = pisi.db.componentdb.ComponentDB()
@@ -86,7 +86,7 @@ def remove(A, ignore_dep = False, ignore_safety = False, autoremove = False):
     ctx.ui.info(_("""The following list of packages will be removed
 in the respective order to satisfy dependencies:
 """) + util.strlist(order))
-    if len(order) > len(A_0):
+    if len(order) > len(A_0) or force_prompt:
         if not ctx.ui.confirm(_('Do you want to continue?')):
             ctx.ui.warning(_('Package removal declined'))
             return False
@@ -107,6 +107,10 @@ in the respective order to satisfy dependencies:
     finally:
         ctx.exec_usysconf()
 
+def remove_orphans(ignore_dep = False, ignore_safety = False):
+    pg, pkgs = plan_autoremove_all()
+    return remove(pkgs, ignore_dep, ignore_safety, autoremove = False, force_prompt=True)
+
 def plan_remove(A):
     # try to construct a pisi graph of packages to
     # install / reinstall
-- 
2.15.1

